_stbt/core.py

```diff
@@ -1191,6 +1191,98 @@ def argparser():
     return parser
 
 
+def _memoize_property_fn(fn):
+    @functools.wraps(fn)
+    def inner(self):
+        # pylint: disable=W0212
+        if fn not in self._FrameObject__frame_object_cache:
+            self._FrameObject__frame_object_cache[fn] = fn(self)
+        return self._FrameObject__frame_object_cache[fn]
+    return inner
+
+
+def _noneify_property_fn(fn):
+    @functools.wraps(fn)
+    def inner(self):
+        if self:
+            return fn(self)
+        else:
+            return None
+    return inner
+
+
+class _FrameObjectMeta(type):
+    def __new__(cls, name, parents, dct):
+        for k, v in dct.iteritems():
+            if isinstance(v, property):
+                if v.fset is not None:
+                    raise Exception(
+                        "FrameObjects must be immutable but this property has "
+                        "a setter")
+                f = v.fget
+                f = _memoize_property_fn(f)
+                if k != 'is_visible' and not k.startswith('_'):
+                    f = _noneify_property_fn(f)
+                dct[k] = property(f)
+
+        return super(_FrameObjectMeta, cls).__new__(cls, name, parents, dct)
+
+    def __init__(cls, name, parents, dct):
+        property_names = sorted([
+            p for p in dir(cls)
+            if isinstance(getattr(cls, p), property)])
+        assert 'is_visible' in property_names
```

Comment:
This is a genuine assertion.  `is_visible` is defined in the base class so this should really never happen.